addon/model/live-mark-set.ts

```diff
@@ -0,0 +1,139 @@
+import Datastore from '@lblod/ember-rdfa-editor/model/util/datastore/datastore';
+import Controller from '@lblod/ember-rdfa-editor/model/controller';
+import EventBus, {
+  AnyEventName,
+  EditorEventListener,
+  ListenerConfig,
+} from '@lblod/ember-rdfa-editor/utils/event-bus';
+import { ModelError } from '@lblod/ember-rdfa-editor/utils/errors';
+import { AttributeSpec } from '@lblod/ember-rdfa-editor/model/mark';
+import { TextMatch } from '@lblod/ember-rdfa-editor/utils/match-text';
+
+export type LiveMarkSpec =
+  | string
+  | {
+      name: string;
+      attributesBuilder?: (textMatch: TextMatch) => AttributeSpec;
+    };
+
+export interface LiveMarkSetArgs {
+  datastoreQuery: (datastore: Datastore) => TextMatch[];
+  liveMarkSpecs: LiveMarkSpec[];
+}
+
+export interface TextMatchDiff {
+  matchesToRemove: Set<TextMatch>;
+  matchesToAdd: Set<TextMatch>;
+}
+
+export default class LiveMarkSet {
+  private _controller: Controller;
+  private _datastoreQuery: (datastore: Datastore) => TextMatch[];
+  private _privateBus: EventBus;
+  private _liveMarkSpecs: LiveMarkSpec[];
+  private _activeMatches: Map<string, TextMatch>;
+
+  constructor(
+    controller: Controller,
+    { datastoreQuery, liveMarkSpecs }: LiveMarkSetArgs
+  ) {
+    this._controller = controller;
+    this._datastoreQuery = datastoreQuery;
+    this._privateBus = new EventBus();
+    this._controller.onEvent('contentChanged', this.update);
+    this._liveMarkSpecs = liveMarkSpecs;
+    this._activeMatches = new Map<string, TextMatch>();
+  }
+
+  get controller(): Controller {
+    return this._controller;
+  }
+
+  get datastoreQuery(): (datastore: Datastore) => TextMatch[] {
+    return this._datastoreQuery;
+  }
+
+  get liveMarkSpecs(): LiveMarkSpec[] {
+    return this._liveMarkSpecs;
+  }
+
+  private update = () => {
+    const { matchesToAdd, matchesToRemove } = this.calculateRanges();
+    const mutator = this.controller.getMutator();
+    const marksRegistry = this.controller.marksRegistry;
+
+    console.log('TOADD', matchesToAdd);
```

Comment:
probably want to use a logger here instead of console.log, or remove them alltogether
```suggestion

```